implement path-based route matching

calebdoxsey opened the pull request:
## Summary
This set of changes adds support for matching routes by path in addition to host. The following 3 fields have been added:

* `prefix`: which matches a URL path based on its prefix (`path: /admin/` matches `/admin/example`)
* `path`: which matches a URL path exactly (`path: /admin` matches `/admin`, but not `/admin/example`)
* `regex`: which matches a URL path via a regular expression. The regular expression syntax is the same as Go's [regexp package](https://pkg.go.dev/regexp?tab=doc).

More restrictive routes should be placed before less restrictive ones, since they're tried in order and the first one to match will go through. ie:

```yaml
policy:
  - from: http://from.example.com
    to: http://to.example.com
    path: /admin
    allowed_domains: ["example.com"]

  - from: http://from.example.com
    to: http://to.example.com
    allow_public_unauthenticated_access: true
```

Both the proxy routing code and the authorization code have been updated to account for these new fields. 

## Related issues
- https://github.com/pomerium/pomerium/issues/378
- https://github.com/pomerium/pomerium/issues/591
- https://github.com/pomerium/pomerium/issues/607


**Checklist**:
- [x] add related issues
- [x] updated docs
- [x] updated unit tests
- [x] updated UPGRADING.md
- [x] ready for review

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/pomerium/pomerium?pullRequest=615) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/pomerium/pomerium?pullRequest=615) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/pomerium/pomerium?pullRequest=615) it.</sub>

calebdoxsey commented:
@desimone here's what I have so far. I plan to add a few more tests, enable the OPA tests in CI and update the documentation, but thought it would be worthwhile to get some feedback, since we may want to change field names, etc.

desimone commented on authorize/evaluator/opa/policy/authz.rego:
```diff
@@ -53,12 +53,67 @@ allow {
+	path = normalize_url_path(rawpath)
+}
+
+normalize_url_path(str) = "/" {
```
🤯

desimone commented on proxy/proxy_test.go:
```diff
@@ -276,3 +276,73 @@ func TestNewReverseProxy(t *testing.T) {
 	}
 }
+
+func TestRouteMatcherFuncFromPolicy(t *testing.T) {
```
😍

desimone replied:
Ahh, maybe add a test for trailing slash on a naked host?

calebdoxsey replied:
I added a test with a trailing slash and it worked.

desimone commented on authorize/evaluator/opa/policy/authz.rego:
```diff
@@ -53,12 +53,67 @@ allow {
+	startswith(input_url_obj.path, policy.prefix)
+}
+
+allowed_route_path(input_url_obj, policy) {
```
does this handle the trailing slash?

calebdoxsey replied:
I added a few more test cases to check. Seems to work as expected.

desimone commented on proxy/proxy.go:
```diff
@@ -323,3 +325,42 @@ func (p *Proxy) roundTripperFromPolicy(policy *config.Policy) http.RoundTripper
+// routeMatcherFuncFromPolicy returns a mux matcher function which compares an http request with a policy.
+//
+// Routes can be filtered by the `source`, `prefix`, `path` and `regex` fields in the policy config.
+func routeMatcherFuncFromPolicy(policy config.Policy) mux.MatcherFunc {
```
really clever function, I like it.

calebdoxsey commented:
@desimone I turned on the OPA tests in CI, updated the configuration documentation, added a more thorough test of routing via prefix, and updated the `upgrading.md` file. 

I think this PR should be ready for a full review for merging at your earliest convenience.

branchmispredictor commented:
> Also, it's unclear to me if there's even a standard way get the forwarded path/uri from a fronting proxy. For instance, I know traefik sets X-Forwarded-Uri (if you [have insecure mode set?](https://community.containo.us/t/forwardauth-does-not-pass-x-forwarded-uri/3277/2) but I don't recall if nginx does, or what other proxies do.

For traefik, it looks like forward auth will set `X-Forwarded-Uri` to the request uri as long as `trustForwardHeader` is not set? [source](https://github.com/containous/traefik/blob/aa21351d0d70546d35634ae33858a93d156766f3/pkg/middlewares/auth/forward.go#L208-L216).

desimone commented:
To circle back on this somewhat, I've found that for `nginx-ingress` the original url is passed as ` X-Original-URL`. What I'm less clear about is whether that's unique to all nginx deployments, or just [ingress-nginx](https://github.com/kubernetes/ingress-nginx/blob/d18fa90cfde32a53166193e5066056aac7267a2b/rootfs/etc/nginx/template/nginx.tmpl#L949). 

<img width="526" alt="Screen Shot 2020-04-18 at 9 53 10 PM" src="https://user-images.githubusercontent.com/1544881/79707023-f6187480-826f-11ea-9b4f-07ab0fab4d9e.png">

calebdoxsey commented:
@desimone I rebased the PR to pull in the latest master changes. Forward auth is working with nginx by passing the URI explicitly:

```yaml
'nginx.ingress.kubernetes.io/auth-url': 'https://forward-authenticate.localhost.pomerium.io/verify?uri=$scheme://$host$request_uri',
'nginx.ingress.kubernetes.io/auth-signin': 'https://forward-authenticate.localhost.pomerium.io/?uri=$scheme://$host$request_uri',
```

The code now handles that case via https://github.com/pomerium/pomerium/pull/615/files#diff-fc1035c92a7c3cbf9faaceb72312de13R118.

I will try testing with traefik tomorrow.

desimone commented on proxy/forward_auth.go:
```diff
@@ -143,3 +141,10 @@ func (p *Proxy) Verify(verifyOnly bool) http.Handler {
 	})
 }
+
+func (p *Proxy) getOriginalRequest(r *http.Request, originalURL *url.URL) *http.Request {
```
Could you point me in the direction of where you first saw this copy-request pattern. I need to get in the habit of this. Would like to better understand when and where it makes sense.

calebdoxsey replied:
It's probably not necessary... but I saw it with the reverse proxy code. https://github.com/golang/go/issues/20068

desimone commented on authorize/evaluator/opa/policy/authz.rego:
```diff
@@ -8,57 +8,120 @@ default allow = false
 	token.payload.email = route_policies[route].allowed_users[_]
 	token.valid
 	count(deny)==0
+	trace(sprintf("allow by email (route=%v email=%v)", [route, token.payload.email]))
```
figure you planned on removing the trace statements, but just in case fyi

beorn- commented:
any date for a 0.7.6 release ? :)

desimone commented:
@beorn- We are shooting to release 0.8.0 on Monday.

fightforlife commented:
I just updated my installation to the latest master branch, (higher than 0.9.1)
Till now my policy looked like this:

```
# personal API
  - from: https://api.domain.com
    to: http://192.168.1.2:9002/api
    allow_public_unauthenticated_access: true
    allow_websockets: true
```

But this is no longer working, since the path in the outgoing url is no longer respected? As far as I undesternd the 3 options in this change only allow to match incoming urls, not outgoing. Is that right?
